File: C++/Heranca/exercicio2.cpp
#include <iostream>
#include <string>

class Animal {
public:
    Animal(std::string nome) : nome(nome) {}

    virtual std::string fazerSom() {
        return "Som genérico do animal";
    }

    std::string getNome() {
        return nome;
    }

private:
    std::string nome;
};

class Cachorro : public Animal {
public:
    Cachorro(std::string nome) : Animal(nome) {}

    std::string fazerSom() override {
        return "Au Au";
    }
};

class Gato : public Animal {
public:
    Gato(std::string nome) : Animal(nome) {}

    std::string fazerSom() override {
        return "Miau";
    }
};

int main() {
    // Teste
    Cachorro cachorro("Flokus");
    Gato gato("Felicete");

    std::cout << "Cachorro faz som: " << cachorro.fazerSom() << "\n";
    std::cout << "Gato faz som: " << gato.fazerSom() << "\n";

    return 0;
}

File: PrimeiraParte/Java/Heranca/exercicio2.java
class Cachorro extends Animal {
    public Cachorro(String nome) {
        super(nome);
    }

    @Override
    public String fazerSom() {
        return "Au Au";
    }
}

class Gato extends Animal {
    public Gato(String nome) {
        super(nome);
    }

    @Override
    public String fazerSom() {
        return "Miau";
    }
}

public class exercicio2 {
    public static void main(String[] args) {
        // Teste
        Cachorro cachorro = new Cachorro("Flokus");
        Gato gato = new Gato("Felicete");

        System.out.println("Cachorro faz som: " + cachorro.fazerSom());
        System.out.println("Gato faz som: " + gato.fazerSom());
    }
}

